static/src/javascripts/projects/commercial/modules/messenger.ts

```diff
@@ -0,0 +1,382 @@
+import reportError from '../../../lib/report-error';
+import { postMessage } from './messenger/post-message';
+
+/**
+ * A message that is sent from an iframe following a standard format
+ *
+ * TODO Is this format formally defined somewhere?
+ */
+type StandardMessage<T = unknown> = {
+	id: string;
+	type: string;
+	iframeId?: string | null | undefined;
+	slotId?: string | null | undefined;
+	/**
+	 * The `value` property is generic since it is up to the sender to attach arbitrary data
+	 *
+	 * We mostly treat this as unknown and leave it up to the message
+	 * listeners to convert to a type they can handle
+	 */
+	value: T;
+};
+
+/**
+ * A legacy from programmatic ads running in friendly iframes. They can
+ * on occasion be larger than the size returned by DFP. And so they
+ * have been setup to send a message of the form:
+ */
+type ProgrammaticMessage = {
+	type: string;
+	value: {
+		id?: string | null | undefined;
+		slotId?: string | null | undefined;
+		height: number;
+		width: number;
+	};
+};
+
+/**
+ * Callbacks that can be registered to fire when receiving messages from an iframe
+ */
+type ListenerCallback = (
+	/**
+	 * The data payload sent by an iframe, and has type `unknown` because we can't
+	 * predict what the iframe will send. It is the responsibility of the callback
+	 * to obtain a value of the desired type (or fail gracefully)
+	 */
+	specs: unknown | null | undefined,
+	/**
+	 * Non-persistent callbacks can be chained together. This value is the return
+	 * value of the previously fired callback in the chain. It is the responsibility
+	 * of the current callback to either ignore it or use it / pass along
+	 */
+	ret: unknown,
+	/**
+	 * Reference to the iframe that is the source of the message
+	 */
+	iframe?: HTMLIFrameElement | null | undefined,
+) => unknown;
+
+/**
+ * The set of listeners currently registered
+ *
+ * Each message sent by an iframe has a `type` field which indicates the kind of
+ * message e.g. `resize`, `measure-ad-load`. One or more listeners is registered
+ * for each type.
+ */
+type Listeners = Record<
+	string,
+	ListenerCallback | ListenerCallback[] | undefined | null
+>;
```

Comment:
I really dislike this. The fact that something can be a single callback or an array of callbacks of the same type is just very confusing.

Why do we need to have `persistent` listeners? What do they mean? Thorny questions…

Reply:
Yup, it's a good point and I'm in agreement with you. I tried to to treat this initially as mostly just adding types and not changing the underlying data structures, but this does seem like the sort of structure you end up with when you're dynamic typing and we can do better in TS.

I don't know what you're feelings are on how you'd change this, some ideas:
- Ditch persistent listeners all together? Can we do this?
- Persistent callbacks as just one-element arrays.
- Convert it so at least these things have tags which we can use to more easily `switch` on e.g.

```typescript
type Listeners = Record<
	string,
	{ kind: 'persistent', callback: ListenerCallback } | 
        { kind: 'non-persistent', callbacks: ListenerCallback[] } |
        undefined
>;
```

Reply:
I still don’t understand what makes a persistent callback different, really.

I would do either `1–ditch` or `3–typed objects` in your list. Ideally, **ditch**!